contracts/StablecoinConverter.sol

```diff
@@ -173,12 +182,62 @@ contract StablecoinConverter is EpochTokenLocker {
         orders[owner][orderId].sellAmount = newSellAmount;
     }
 
-    function writeCurrentPrices(
+    function updateCurrentPrices(
         uint128[] memory prices,  //list of prices for touched token only
         uint16[] memory tokenIdsForPrice  // price[i] is the price for the token with tokenID tokenIdsForPrice[i]
     ) internal {
+        for (uint i = 0; i < previousSolutionTokenIdForPrice.length; i++) {
+            currentPrices[previousSolutionTokenIdForPrice[i]] = 0;
+        }
         for (uint i = 0; i < tokenIdsForPrice.length; i++) {
             currentPrices[tokenIdsForPrice[i]] = prices[i];
         }
     }
+
+    function documentTrades(
+        uint batchIndex,
+        address[] memory owners,  //tradeData is submitted as arrays
+        uint16[] memory orderIds,
+        uint128[] memory volumes,
+        uint128[] memory prices,
+        uint16[] memory tokenIdsForPrice
+    ) internal {
+        previousSolutionBatchId = batchIndex;
+        for (uint i = 0; i < owners.length; i++) {
+            previousSolutionTrades.push(TradeData({
+                owner: owners[i],
+                orderId: orderIds[i],
+                volume: volumes[i]
+            }));
+        }
+        for (uint i = 0; i < prices.length; i++) {
+            previousSolutionTokenIdForPrice.push(tokenIdsForPrice[i]);
+        }
+    }
+
+    function undoPreviousSolution(uint batchIndex) internal {
+        if (previousSolutionBatchId == batchIndex) {
+            for (uint i = 0; i < previousSolutionTrades.length; i++) {
+                address owner = previousSolutionTrades[i].owner;
+                uint orderId = previousSolutionTrades[i].orderId;
+                Order memory order = orders[owner][orderId];
+                uint sellVolume = previousSolutionTrades[i].volume;
+                addBalance(owner, tokenIdToAddressMap(order.sellToken), sellVolume);
+            }
+            for (uint i = 0; i < previousSolutionTrades.length; i++) {
+                address owner = previousSolutionTrades[i].owner;
+                uint orderId = previousSolutionTrades[i].orderId;
+                Order memory order = orders[owner][orderId];
+                uint sellVolume = previousSolutionTrades[i].volume;
+                uint buyVolume = sellVolume
```

Comment:
Does this logic already exist somewhere else and can be refactored in a common method?

Reply:
I know you ask these questions rhetorically :D
However, I currently see only the calculation for sellVolume and buyVolume as refactorable. But is it worth doing so, if these are just one single line each?

Reply:
I think any complex logic (even if it's just one line) can be refactored in a method. It's about having this logic in one place in case we need to change it later.